src/renderer/state.ts

```diff
@@ -654,7 +696,7 @@ export class AppState {
 
   public addNewVersions(versions: RunnableVersion[]) {
     for (const ver of versions) {
-      this.versions[ver.version] ||= ver;
+      this.versions[ver.version] = ver;
```

Comment:
I'm not a fan of this behavior change as it no longer matches the name of the method, `addNewVersions`.

Reply:
A couple of alternatives I can think of:

- instead of calling `addNewVersions` in the `message` listener, it would be simple enough to just update the version in `versions` directly:
```ts
this.versions[payload.version] = payload;
```

- rename `addNewVersions` to something like `addOrUpdateVersions` and, if you're also concerned about always replacing the current version vs. only ever adding values that don't exist yet, we could update the existing version only if some of the properties we try to update in this PR is different:
```ts
public addOrUpdateVersions(versions: RunnableVersion[]) {
  for (const ver of versions) {
    const currentVersion = this.versions[ver.version];

    if (
      !currentVersion ||
      currentVersion.downloadProgress !== ver.downloadProgress ||
      currentVersion.state !== ver.state
    ) {
      this.versions[ver.version] = ver;
    }
  }
}
```

I think the first one is the least disrupting and it definitely makes for a shorter diff, but I'm open to suggestions.

Reply:
Per my other comment about making the broadcast message more generic and taking an array of versions, how about we add a new method, `setVersionStates(versions: RunnableVersion[])` which contains the original change here the comment is on - just a for loop and sets the state.

Reply:
Just a sanity check to see if I'm still following:

> add a new method, setVersionStates(versions: RunnableVersion[]) which contains the original change here the comment is on - just a for loop and sets the state.

If I understand correctly, we'd replace the `updateVersionState` method we currently have with:
```ts
private setVersionStates(versions: RunnableVersion[]) {
  for (const ver of versions) {
    this.versions[ver.version] = ver;
  }
}
```

> Make a private AppState.addNewVersionsImpl and move the existing code there

That would be just another name for the `setVersionStates` method I just mentioned above, right?

> Then broadcast the message in AppState.addNewVersions with the full list of versions
```ts
public addNewVersions(versions: RunnableVersion[]) {
  for (const ver of versions) {
    this.versions[ver.version] ||= ver;
  }

  this.broadcastChannel.postMessage({
    type: AppStateBroadcastMessageType.syncVersions,
    payload: versions,
  });
}

// broadcast listener
switch (type) {
  case AppStateBroadcastMessageType.syncVersions: {
    this.setVersionStates(payload);

    break;
  }

  // ...
}
```

If this snippet is right, it would involve replacing our current `this.updateVersionState(ver)` calls with `this.addNewVersions([ver])` to update the version state across windows, since the `addNewVersions` method is the only place where we'd actually be broadcasting the message - and it feels odd to call `addNewVersions` when updating the download progress, for example.

Reply:
You're right, things have shifted a little since my original comment. This is the overall sketch I have in my head now, with a slight rename to avoid confusion:
* Rename the existing (in this PR) `updateVersionState` method to `broadcastVersionStates(versions: RunnableVersion[])`
* `setVersionStates(versions: RunnableVersion[])` is a new method all together
* `addNewVersions` would also use `broadcastVersionStates`
* The broadcast listener uses `setVersionStates` as you have above

I *think* that covers all the states.

Reply:
Thank you, that's clear now and it works fine. Before pushing that change, I'd like to take a step back to your original comment:

> We should probably handle the case where one window updates the Electron versions (can happen manually under settings, or by loading a gist with a new version specified)

If I understand the requirements correctly, we can also do that by simply rolling back `addNewVersions` to what it was before this PR (`this.versions[ver.version] ||= ver`) and using `this.versions[payload.version] = payload` in the broadcast listener as I suggested in this thread. As you can see in the video below, if the left window downloads v25.0.0, the right window will be updated as expected:

https://github.com/electron/fiddle/assets/19478575/7cb912aa-06e6-40a0-a1ab-e29eafe6e4cc

I built Fiddle locally and tried to open a fiddle from the docs manually from the command line, and AFAICT, it works as expected — when the fiddle is loaded in one window, the other window will reflect any changes to the version the fiddle uses:

https://github.com/electron/fiddle/assets/19478575/bad946e2-0e8c-45ae-ae17-5a7cc711359c

I'm double-checking because that seems a lot simpler, but I'm happy to toss this approach if there are any downsides or corner cases I'm missing.

(As an aside, notice in the second video that the version selector in the second window highligths v25.2.0 when I load the fiddle in the first window. The component probably uses the version in localStorage to determine which version is active, and changing that would involve some more thinking, probably worth another issue).

Reply:
I removed the `releases.json` file from my AppData folder (and also had to remove one of the versions from the `./static/releases.json` project file — if the file is missing or is an empty array, Fiddle won't start correctly), and yeah my approach fails miserably in that case 😅.

The approach you suggested does work. At first, I tried to only leave one version in my `./static/releases.json` file, but it took ~33 _seconds_ for the other window to update its state after receiving the broadcast with ~1900 versions and start responding to user interaction again — this scenario sounds very unlikely, so we should be safe when updating even a few dozen versions.

Reply:
> The approach you suggested does work. At first, I tried to only leave one version in my ./static/releases.json file, but it took ~33 seconds for the other window to update its state after receiving the broadcast with ~1900 versions and start responding to user interaction again — this scenario sounds very unlikely, so we should be safe when updating even a few dozen versions.

It looks like that's because `setVersionStates` is triggering mobx to cause updates on every version change. We can fix this by wrapping it with `runInAction` so that it's a single update instead of thousands. Tested locally and it resolves this issue - see my suggested change.

Reply:
I went ahead and marked `setVersionStates` as an `action` in the `makeObservable` call in the constructor, so we don't end up shooting ourselves in the foot by forgetting to wrap the calls to that method with `runInAction` 😅